mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp

```diff
@@ -444,10 +452,63 @@ static bool isUnfusedKernelStore(rock::GlobalStoreOp store) {
   return ret;
 }
 
+LogicalResult
+ReduceRewritePattern::matchAndRewrite(rock::ReduceOp reduceOp,
+                                      PatternRewriter &rewriter) const {
+  Location loc = reduceOp.getLoc();
+  GlobalStoreOp globalStoreOp = traceToGlobalStore(reduceOp.getIn());
+  if (reduceOp.getReduceMethod() != ReduceMethod::Sum) {
+    return reduceOp.emitError("We only support sum reductions.!");
+  }
+  if (!globalStoreOp) {
+    return reduceOp.emitError(
+        "rock.reduce input is not the global store of the previous op.");
+  }
+  int64_t elementCount = globalStoreOp.getLengthAttr().getInt();
+  Type elementType = globalStoreOp.getSource().getType().getElementType();
+  Value zero = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+  int64_t reductionAxis = reduceOp.getAxisAttr().getInt();
+  int64_t inpRank = reduceOp.getIn().getType().cast<ShapedType>().getRank();
+  if (elementCount != 1 && reductionAxis == inpRank - 1) {
+    // if the reduction dimension is the last dimension and its a vector.
+    OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPoint(globalStoreOp);
+    VectorType loadedElType = VectorType::get({elementCount}, elementType);
+    Value loadVal = rewriter.create<vector::LoadOp>(
```

Comment:
I don't think this is correct? Consider the possibility that the vector length is 4 and the output is a NxKx2x2 tensor. Adding each thread's values together will lead to too much reduction.

Reply:
right! yes -- vector length exceeding product of lowest dimSizes.. I ll add a test and a fix.

Reply:
I wouldn't waste too much time. In the fusion refactor (which is a few days out, hopefully - or at least it feels like it's going places) - the vectorization length won't be decided yet when we get to align_tiling.